### PR TITLE
EPMLSTRCMW-254 Feat: Add getFiles in ProjectFileController

### PR DESCRIPTION
**Pull request by DruzhininPavel**
**DESCRIPTION:**
add getFiles on ProjectFileService
change endpoint getFile to the REST-full
change endpoint buildConfiguration to the REST-full

CHANGES:
ProjectFileService - added getFiles method
ProjectFileController - added getFiles method and chanched endpoints from Queru to Segment parameters
ProjectFileService - just fixef formatting
AggregationService - getFilesByProjectId method have been rewritten
GitLabProjectVersioning - added getFiles method
ProjectVersioning - added getFiles method
ProjectFileServiceTest - added 3 tests
ProjectFileControllerTest - added 3 tests, changed 6 tests
AggregationServiceTest - several tests have been rewritten
GitLabProjectVersioningTest - added 2 tests

### CODE REVIEW

**Review by OlegKoshelev — APPROVED** [review]
well done!

File: services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala [hunk 1]
```diff
@@ -126,7 +128,25 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
         .recover { case e: Throwable => Left(VersioningException.HttpException(e.getMessage)) }
     }
 
-  override def getFiles(project: Project, path: Path)(implicit ec: ExecutionContext): AsyncResult[List[String]] = ???
+  override def getFiles(
```
Comment (Nikolai-Isaev) [inline services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:129]: Maybe it will be better to place it near the getFile method

**Review by Nikolai-Isaev — APPROVED** [review]
Lgtm

File: services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala (hunk 1 above)
Comment (DruzhininPavel) [reply services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:129]: Possibly, but it will cause more changes to the file and force the reviewer to do more work.

File: controllers/src/main/scala/cromwell/pipeline/controller/utils/CustomPathMatcher.scala [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package cromwell.pipeline.controller.utils
+
+import akka.http.scaladsl.server
+import akka.http.scaladsl.server.PathMatchers.Segment
+import cromwell.pipeline.datastorage.dto
+
+import java.nio.file.{ Path, Paths }
+
+object CustomPathMatcher {
```
Comment (myazinn) [inline controllers/src/main/scala/cromwell/pipeline/controller/utils/CustomPathMatcher.scala]:
```suggestion
object PathMatchers {
```
It's obvious that these path matchers are custom, otherwise we wouldn't have defined them at all :) So we can just make the name similar to predefined matchers

File: controllers/src/test/scala/cromwell/pipeline/controller/ProjectFileControllerTest.scala [hunk 3]
```diff
@@ -51,7 +52,7 @@ class ProjectFileControllerTest extends AsyncWordSpec with Matchers with Scalate
     }
 
     "upload file" should {
-      val projectFile = ProjectFile(Paths.get("folder/test.txt"), projectFileContent)
+      val projectFile = ProjectFile(Paths.get(pathString), projectFileContent)
```
Comment (myazinn) [inline controllers/src/test/scala/cromwell/pipeline/controller/ProjectFileControllerTest.scala]:
```suggestion
      val projectFile = ProjectFile(path, projectFileContent)
```

File: services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala [hunk 4]
```diff
@@ -22,9 +23,9 @@ trait ProjectVersioning[E >: VersioningException] {
     implicit ec: ExecutionContext
   ): AsyncResult[Project]
 
-  def getFiles(project: Project, path: Path)(
+  def getFiles(project: Project, version: Option[PipelineVersion] = None)(
     implicit ec: ExecutionContext
-  ): AsyncResult[List[String]]
+  ): EitherT[Future, VersioningException, List[ProjectFile]]
```
Comment (myazinn) [inline services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala]: Why did you change return type from AsyncResult to EitherT? Now it doesn't match with signature of other methods that return AsyncResult

File: services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala (hunk 1 above)
Comment (myazinn) [inline services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:129]: Looks like now this method can be re-used in `cromwell.pipeline.service.AggregationService#getFilesByProjectId`

**Review by myazinn — CHANGES_REQUESTED** [review]

File: controllers/src/main/scala/cromwell/pipeline/controller/utils/CustomPathMatcher.scala (hunk 2 above)
Comment (DruzhininPavel) [reply controllers/src/main/scala/cromwell/pipeline/controller/utils/CustomPathMatcher.scala]: Ok. I Agree ;) I`ll fix it.

File: controllers/src/test/scala/cromwell/pipeline/controller/ProjectFileControllerTest.scala (hunk 3 above)
Comment (DruzhininPavel) [reply controllers/src/test/scala/cromwell/pipeline/controller/ProjectFileControllerTest.scala]: Yep, you are right. It will be changed.

File: services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala (hunk 1 above)
Comment (DruzhininPavel) [reply services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala:129]: Done.

File: services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala (hunk 4 above)
Comment (DruzhininPavel) [reply services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala]: We need to discuss it.

File: controllers/src/main/scala/cromwell/pipeline/controller/utils/CustomPathMatcher.scala (hunk 2 above)
Comment (DruzhininPavel) [reply controllers/src/main/scala/cromwell/pipeline/controller/utils/CustomPathMatcher.scala]: Done.

File: controllers/src/test/scala/cromwell/pipeline/controller/ProjectFileControllerTest.scala (hunk 3 above)
Comment (DruzhininPavel) [reply controllers/src/test/scala/cromwell/pipeline/controller/ProjectFileControllerTest.scala]: Done.

File: services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala (hunk 4 above)
Comment (DruzhininPavel) [reply services/src/main/scala/cromwell/pipeline/service/ProjectVersioning.scala]: Done

**Review by gominasian — APPROVED** [review]
LGTM

File: services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala [hunk 5]
```diff
@@ -361,6 +361,73 @@ class GitLabProjectVersioningTest extends AsyncWordSpec with Matchers with Mocki
       }
     }
 
+    "getFiles" should {
+      val dummyFilesTree: List[FileTree] = List(dummyFileTree)
+      val path = Paths.get(dummyFilesTree.head.path)
```
Comment (myazinn) [inline services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala]:
```suggestion
      val path = Paths.get(dummyFileTree.path)
```

File: services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala [hunk 6]
```diff
@@ -361,6 +361,73 @@ class GitLabProjectVersioningTest extends AsyncWordSpec with Matchers with Mocki
...
+      val encodedPathStr = URLEncoderUtils.encode(path.toString)
+      //        val emptyListValidateResponse = "List((,List(JsonValidationError(List(error.expected.jsarray),WrappedArray())))"
```
Comment (myazinn) [inline services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala]: Why do we need it? Never commit an unused commented code

File: services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala [hunk 7]
```diff
@@ -361,6 +361,73 @@ class GitLabProjectVersioningTest extends AsyncWordSpec with Matchers with Mocki
...
+      val version = dummyPipelineVersion
+
+      def request(project: Project, version: Option[PipelineVersion]): Future[Response[List[FileTree]]] = {
```
Comment (myazinn) [inline services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala]: Let's give it a better name. Something like `getFileTrees` should be good enough

**Review by myazinn — CHANGES_REQUESTED** [review]
Changes look good to me except for few small issues. However, you should also modify postman collection according to your API changes

File: services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala (hunk 5 above)
Comment (DruzhininPavel) [reply services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala]: Done.

File: services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala (hunk 6 above)
Comment (DruzhininPavel) [reply services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala]: Yep, my fault. Fixed.

File: services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala (hunk 7 above)
Comment (DruzhininPavel) [reply services/src/test/scala/cromwell/pipeline/service/GitLabProjectVersioningTest.scala]: Renamed all requests to clearer names.

File: postman/Files.postman_collection.json [hunk 8]
```diff
@@ -5,6 +5,92 @@
...
+			"request": {
+				"method": "GET",
+				"header": [
+					{
+						"key": "Authorization",
+						"value": "{{auth_token}}",
+						"type": "text"
+					}
+				],
+				"body": {
```
Comment (myazinn) [inline postman/Files.postman_collection.json]: Why do we need empty body in GET request?

File: postman/Files.postman_collection.json [hunk 9]
```diff
@@ -5,6 +5,92 @@
...
```
Comment (myazinn) [inline postman/Files.postman_collection.json]: same here

**Review by myazinn — APPROVED** [review]

**Review by Dmitrii-Shevchenko — APPROVED** [review]
